sdcm/utils/k8s.py

```diff
@@ -609,6 +614,251 @@ def stop(self, timeout=None):
+            labels = metadata.get('labels', {})
+            if not all((key in labels.keys() for key in self.SCYLLA_PODS_EXPECTED_LABEL_KEYS)):
+                return None
+
+            if name := metadata.get('name'):
+                if name not in self.mapper_dict[namespace]:
+                    self.mapper_dict[namespace][name] = {}
+                if 'current_ip' not in self.mapper_dict[namespace][name]:
+                    self.mapper_dict[namespace][name]['current_ip'] = None
+                if 'old_ips' not in self.mapper_dict[namespace][name]:
+                    self.mapper_dict[namespace][name]['old_ips'] = []
+            for subset in data.get('object', {}).get('subsets', []):
+                for address in subset.get('addresses', []):
+                    current_ip_address = address['ip']
+                    if current_ip_address == self.mapper_dict[namespace][name]['current_ip']:
+                        continue
+                    old_ip_candidate = self.mapper_dict[namespace][name]['current_ip']
+                    self.mapper_dict[namespace][name]['current_ip'] = current_ip_address
+                    if not old_ip_candidate:
+                        break
+                    self.mapper_dict[namespace][name]['old_ips'].append(old_ip_candidate)
+                    LOGGER.info(
+                        "'%s/%s' node has changed it's pod IP address from '%s' to '%s'",
+                        namespace, name, old_ip_candidate, current_ip_address)
+                    self._call_callbacks(namespace, name)
+                    break
+                else:
+                    break
+        except Exception as exc:  # pylint: disable=broad-except
+            LOGGER.warning(
+                "Failed to parse following line: %s\nerr: %s", line, exc)
+
+    def _call_callbacks(self, namespace, pod_name) -> None:
+        # TODO: run callbacks in parallel on per-namespace basis and serially inside a
+        #       single namespace.
+
+        def process_callback(callback, namespace, pod_name=None, add_pod_name_as_kwarg=False):
+            suffix = f"(namespace={namespace}, pod_name={pod_name})"
+            func = callback[0]
+            args = callback[1]
+            kwargs = {} | callback[2]
+            if add_pod_name_as_kwarg:
+                kwargs['pod_name'] = pod_name
+            # TODO: make following be 'debug' message
+            LOGGER.info("Calling '%s' callback %s", func.__name__, suffix)
+            try:
+                func(*args, **kwargs)
+            except Exception as exc:  # pylint: disable=broad-except
+                LOGGER.warning("Callback call failed %s: %s", suffix, str(exc))
+
+        data = self.mapper_dict.get(namespace, {})
+        for dict_key in (pod_name, '__each__'):
+            callbacks = data.get(dict_key, {}).get('callbacks', [])
+            if dict_key == '__each__':
+                # NOTE: when 'add_pod_name_as_kwarg' is True it means
```

Comment:
why the order is important ?

Reply:
Because we must refresh IP addresses before reconfiguring the monitoring. If we do it vice-versa then the monitoring update becomes nonsense.

Reply:
o.k.